### PR TITLE
new lint: `manual_c_str_literals`

### PR DESCRIPTION
**Pull request by y21**
With rust-lang/rust#117472 merged and `c""` syntax stabilized, I think it'd be nice to have a lint for using `CStr::from_ptr` (and similar constructors) with a string literal as an argument.
We can probably also lint `"foo\0".as_ptr()` and suggest `c"foo".as_ptr()`. I might add that to this PR tomorrow if I find the time.

The byte string literal to c string literal rewriting is ugly but oh well.

changelog: new lint: `manual_c_str_literals`
[#11919](https://github.com/rust-lang/rust-clippy/pull/11919)

### CODE REVIEW

**Comment by y21** [comment]
Alright, I think it should be ready for review now.

It now also lints bare `"foo\0".as_ptr()` (without a `CStr::from_*` call), inspired by rust-lang/rust#118566. This involved a little bit of extra logic because it shouldn't lint twice for something like `CStr::from_ptr(b"f\0".as_ptr())` (once for the `CStr::from_ptr` expression and the `.as_ptr()` call).

This still needs `#![feature(c_str_literals)]`s in a few places but I think I'll be able to remove them after the next sync when the pinned nightly is bumped.

File: clippy_lints/src/methods/manual_c_str_literals.rs
Comment (y21) [reply clippy_lints/src/methods/manual_c_str_literals.rs:20]: oh you're right, good catch, thanks!

**Comment by y21** [comment]
Do you think that we should wait with merging until `c""` is *actually* usable on the stable channel (when 1.76 gets released)?
Because otherwise, nightly users will get this lint and changing it means that it won't build on the stable channel. 🤔

File: clippy_lints/src/methods/manual_c_str_literals.rs
Comment (xFrednet) [inline clippy_lints/src/methods/manual_c_str_literals.rs:148]: Indexing into a string can be a bit iffy. I believe this is safe, but could you still add a test with multi-byte-chars, like this:

```rust
CStr::from_bytes_with_nul(b"电脑");
CStr::from_bytes_with_nul(b"电脑\0");
CStr::from_bytes_with_nul(b"电脑\x00");
```

**Review by xFrednet — APPROVED** [review]
Tiny nit, and then everything should be good to go

File: clippy_lints/src/methods/manual_c_str_literals.rs
Comment (y21) [reply clippy_lints/src/methods/manual_c_str_literals.rs:148]: Byte string literals can't have non-ascii characters like this, but we lint on `"电脑\0".as_ptr()` too so I used that for the tests (should test the same thing)

**Review by xFrednet — APPROVED** [review]
LGTM, now we're just waiting for the next release :)

**Comment by y21** [comment]
Wouldn't we have to wait until 1.76, which is in two releases, since that's when `c""` literals will be on stable? Or rather, ~two weeks before the 1.76 release, so that it makes it into nightly with the sync just in time.

If we merge when 1.75 gets released, we'd still have the problem that stable cannot yet use that syntax and nightly users likely don't want this lint because they still want their project to build on stable, right?

It's possible that I have the release cycle mixed up though

**Comment by y21** [comment]
Going to hold off fixing conflicts since we're gonna need to wait a bit anyway

**Comment by xFrednet** [comment]
> If we merge when 1.75 gets released, we'd still have the problem that stable cannot yet use that syntax and nightly users likely don't want this lint because they still want their project to build on stable, right?

I'm a bit undecided on this one. It would be cool, to have a lint that suggests the new `c""` literals on stable, one version after the stabilization of the feature. I think it would be better to merge it with 1.75. Alternatively, we could also flag this for a beta backport :thinking:. Then it would be on beta at the same time as the new feature, while only blocking the nightly + stable users for a short time.

Actually, thinking about it, users can just set the minimal rust version, if they want to prevent the lint from triggering on nightly. With that config, I would actually say it's safe to merge it as is.

**Comment by y21** [comment]
Looks like there's an ongoing discussion in the T-lang zulip for reverting the stabilization of c string literals, at least for 1.76
https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/rfc.203349.3A.20mixed.20utf8.20literals/near/409504826

I think if it does end up being reverted, we should probably really wait, since then the MSRV is also wrong. But I do agree with your last point, that if the MSRV is correct, users can set that to not get the lint

**Comment by y21** [comment]
So, C-String literals will be stable in 1.77 and 1.76 will be stable in a few days. Do you think we should try to get this merged soon, or wait?

**Comment by xFrednet** [comment]
Let's merge this soon, if we merge it now, it should be on stable in 1.78.

For some context:
* Version 1.76 is now on the `stable` branch in the `rust` repo
* Version 1.77 is now on the `beta` branch in the `rust` repo
* The next sync will be merged on to master, which will be branched to beta for 1.78

**Comment by y21** [comment]
Squashed the commits and rebased on to master

File: clippy_config/src/msrvs.rs
Comment (xFrednet) [inline clippy_config/src/msrvs.rs:20]: It looks like they're planning to stabilize it in 1.77: 

> We discussed this in the triage call today and agreed, given that https://github.com/rust-lang/rust/pull/119172 has landed, that we were happy to see this restabilize in Rust 1.77
>
> https://github.com/rust-lang/rust/issues/105723#issuecomment-1919576337

On nightly these strings already seem to work, so I think it's safe to merge this, once the version is updated :)

Comment (y21) [reply clippy_config/src/msrvs.rs:20]: oh right, oops, of course 😅

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Perfect, looks good to me, thank you for all the work you put into this. Let's get this on to master:

@bors r+

**Comment by xFrednet** [comment]
MacOS is not our friend rn. We just have to retry until it works.

@bors retry

@bors d+

**Pull request merged**
